### PR TITLE
Add development quick start instructions

### PR DESCRIPTION
**Pull request by jereksel**
Makes forking from https://github.com/arrow-kt/arrow/issues/1742 absolete.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -52,18 +52,36 @@
...
+- compiler-plugin: Kotlin compiler plugin
+- gradle-plugin: Plugin for Gradle for activating Kotlin compiler plugin
+- idea-plugin: Plugin for Intellij IDEA for IDE integration (icons, autocomplete etc.)
+- testing-plugin: Tools for testing compiler and idea plugin
 
-## Credits
+Quick start development environment:
+
+- Run `publishAndRunIde` Gradle task
+- Import sample project (https://github.com/jereksel/arrow-meta-example) in new IDE instance
```
Comment (jereksel) [inline README.md]: Feel free to fork it

Comment (jereksel) [inline README.md]: Feel free to fork it

Comment (jereksel) [reply README.md]: https://github.com/arrow-kt/arrow-meta/pull/2

Comment (raulraja) [reply README.md]: @jereksel you have been invited :) feel free to port the sample project over to a module. Welcome to Meta!

**Comment by raulraja** [comment]
@jereksel ready to go once conflicts are resolved, thanks!

File: build.gradle [hunk 2]
```diff
@@ -21,15 +20,25 @@ allprojects {
...
+  dependsOn ':compiler-plugin:publishToMavenLocal'
   dependsOn ':gradle-plugin:publishArrowPluginMarkerMavenPublicationToMavenLocal'
   dependsOn ':gradle-plugin:publishPluginMavenPublicationToMavenLocal'
-  tasks.findByPath(':gradle-plugin:publishArrowPluginMarkerMavenPublicationToMavenLocal').mustRunAfter 'build'
-  tasks.findByPath(':gradle-plugin:publishPluginMavenPublicationToMavenLocal').mustRunAfter 'build'
+  dependsOn ':gradle-plugin:publishToMavenLocal'
+}
+
+task buildMeta {
+  dependsOn 'build'
```
Comment (i-walker) [inline build.gradle]: Can we add `clean`, too?

File: README.md [hunk 3]
```diff
@@ -52,18 +52,36 @@
 
 ## Contributing
 
-## License
+Arrow-meta project is split into 4 packages:
+- compiler-plugin: Kotlin compiler plugin
+- gradle-plugin: Plugin for Gradle for activating Kotlin compiler plugin
+- idea-plugin: Plugin for Intellij IDEA for IDE integration (icons, autocomplete etc.)
+- testing-plugin: Tools for testing compiler and idea plugin
```
Comment (i-walker) [inline README.md]: Currently, the dsl to create idea plugin tests is in `idea-plugin`.
But is also not merged into master, yet 😄. 
We heavily rely dependencies which we already get from the existing set up in the `idea-plugin`.

File: compiler-plugin/build.gradle [hunk 4]
```diff
@@ -41,6 +44,9 @@ shadowJar {
     configurations = [project.configurations.compile]
     relocate 'org.jetbrains.kotlin.com.intellij', 'com.intellij'
     relocate 'org.jetbrains.kotlin.load', 'kotlin.reflect.jvm.internal.impl.load'
+
+    //Required for idea-plugin tests
+    relocate 'org.apache.log4j', 'shadow.org.apache.log4j'
```
Comment (i-walker) [inline compiler-plugin/build.gradle]: Thank you 😄 🎉

Comment (i-walker) [reply compiler-plugin/build.gradle]: @jansorg Is that the correct dependency? For running it in CI

File: gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowGradlePlugin.kt [hunk 5]
```diff
@@ -20,10 +20,5 @@ class ArrowGradlePlugin : Plugin<Project> {
 
   override fun apply(project: Project): Unit {
     project.extensions.create("arrow", ArrowExtension::class.java)
```
Comment (i-walker) [inline gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowGradlePlugin.kt]: You should not delete that. Otherways we can not add the compiler-plugin to the repo. Once we deploy the compiler plugin we can change that to add the dependency. But as for now, we would keep it.

File: gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowKotlinGradleSubplugin.kt [hunk 6]
```diff
@@ -13,7 +13,7 @@ import org.jetbrains.kotlin.gradle.plugin.SubpluginOption
  */
 class ArrowKotlinGradleSubplugin : KotlinGradleSubplugin<AbstractCompile> {
   companion object {
-    private const val GRADLE_ARTIFACT_NAME = "gradle-plugin"
+    private const val KOTLIN_ARTIFACT_NAME = "compiler-plugin"
```
Comment (i-walker) [inline gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowKotlinGradleSubplugin.kt]: You should not change the names here

File: gradle/wrapper/gradle-wrapper.properties [hunk 7]
```diff
@@ -1,5 +1,6 @@
-distributionUrl=https\://services.gradle.org/distributions/gradle-5.0-bin.zip
-distributionPath=wrapper/dists
+#Sat Oct 26 17:44:20 CEST 2019
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.6.3-all.zip
```
Comment (i-walker) [inline gradle/wrapper/gradle-wrapper.properties]: Can you revert those changes here?

File: gradle-plugin/build.gradle [hunk 8]
```diff
@@ -6,16 +6,13 @@ plugins {
...
-project.group = 'io.arrow-kt'
-project.version = '0.0.1'
 def pluginId = "arrow.meta.plugin.gradle"
 
 dependencies {
   implementation gradleApi()
   implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
   implementation "org.jetbrains.kotlin:kotlin-gradle-plugin-api:$kotlin_version"
   implementation "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
-  compileOnly project(":compiler-plugin") { "embeddable" }
```
Comment (i-walker) [inline gradle-plugin/build.gradle]: Why did you delete this? 😄

File: compiler-plugin/build.gradle [hunk 9]
```diff
@@ -1,5 +1,8 @@
+plugins {
+    id 'java-gradle-plugin'
```
Comment (i-walker) [inline compiler-plugin/build.gradle]: Why do we need those plugins? :)

File: gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowKotlinGradleSubplugin.kt [hunk 10]
```diff
@@ -28,11 +28,19 @@ class ArrowKotlinGradleSubplugin : KotlinGradleSubplugin<AbstractCompile> {
     variantData: Any?,
     androidProjectHandler: Any?,
     kotlinCompilation: KotlinCompilation<KotlinCommonOptions>?
-  ): List<SubpluginOption> = emptyList()
+  ): List<SubpluginOption> {
+
+    @Suppress("UNUSED_VARIABLE")
+    val extension = project.extensions.findByType(ArrowExtension::class.java)
```
Comment (i-walker) [inline gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowKotlinGradleSubplugin.kt]: I can't really tell from the code for what this is used. Can you help me, please ? @jereksel

File: idea-plugin/src/main/kotlin/arrow/meta/plugin/idea/gradle/MetaClasspathContributor.kt [hunk 11]
```diff
@@ -9,6 +9,11 @@ import java.io.File
 import java.io.FileInputStream
 import java.util.jar.JarInputStream
 
+/**
+ * Source:
+ * https://github.com/fabianishere/kotlin-plugin-generated/commit/19fbbe9d222126b33cfe90b604473ba7e7ea371b#diff-2f347766958584e8e50469cc6bab68c9
+ * https://github.com/JetBrains/kotlin/blob/822b455/plugins/kotlin-serialization/kotlin-serialization-ide/src/org/jetbrains/kotlinx/serialization/idea/KotlinSerializationImportHandler.kt
```
Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/plugin/idea/gradle/MetaClasspathContributor.kt]: This link seems to be dead?

File: build.gradle (hunk 2 above)
Comment (jereksel) [reply build.gradle]: I don't like `clean`, because it makes build much longer. I understand using it in specific times like changing branch, or updating gradle - but not during every build.

File: gradle-plugin/build.gradle (hunk 8 above)
Comment (jereksel) [reply gradle-plugin/build.gradle]: Gradle plugin does not need compiler-plugin embedded - it's downloaded from repo

File: gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowKotlinGradleSubplugin.kt (hunk 6 above)
Comment (jereksel) [reply gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowKotlinGradleSubplugin.kt]: For me `GRADLE_ARTIFACT_NAME` means gradle plugin, not kotlin compiler plugin that it's supposed to be

File: gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowKotlinGradleSubplugin.kt (hunk 10 above)
Comment (jereksel) [reply gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowKotlinGradleSubplugin.kt]: When we will be adding options for gradle plugin we must pass them to compiler plugin. It's done in this method

File: idea-plugin/src/main/kotlin/arrow/meta/plugin/idea/gradle/MetaClasspathContributor.kt (hunk 11 above)
Comment (jereksel) [reply idea-plugin/src/main/kotlin/arrow/meta/plugin/idea/gradle/MetaClasspathContributor.kt]: Both works fine here

File: compiler-plugin/build.gradle (hunk 9 above)
Comment (jereksel) [reply compiler-plugin/build.gradle]: Yeah, java-gradle-plugin is not needed, but maven-plugin is required for local maven deployment

File: gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowGradlePlugin.kt (hunk 5 above)
Comment (jereksel) [reply gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowGradlePlugin.kt]: Because of that to use plugin you must clone arrow-meta in specific directory. Gradle plugin has mechanism of automatically downloading and applying kotlin compiler plugin from maven repository.

File: gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowKotlinGradleSubplugin.kt (hunk 10 above)
Comment (i-walker) [reply gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowKotlinGradleSubplugin.kt]: But there are no Options here now. So why add a variable?

File: gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowKotlinGradleSubplugin.kt (hunk 6 above)
Comment (i-walker) [reply gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowKotlinGradleSubplugin.kt]: That is not the intent what you might think.
We supply tooling in the idea plugin to get the gradle-plugin, which does the registration automatically, based on rules we define in `ArrowGradlePlugin`. Because you deleted it it makes sense just to retrieve it directly. 
With the changes, you introduce your naming the gradle-plugin `compiler-plugin`, which doesn't make sense to me.

File: build.gradle (hunk 2 above)
Comment (i-walker) [reply build.gradle]: CI needs it. I don't mind having it on your settings, but that introduces changes to our process.

File: gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowGradlePlugin.kt (hunk 5 above)
Comment (i-walker) [reply gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowGradlePlugin.kt]: Can you explain how that works. We did not have that experience that Gradle does that?

File: compiler-plugin/build.gradle (hunk 9 above)
Comment (i-walker) [reply compiler-plugin/build.gradle]: Nop it is not.

File: compiler-plugin/build.gradle (hunk 4 above)
Comment (jansorg) [reply compiler-plugin/build.gradle]: @i-walker The plugin is using the logging facility of the IntelliJ SDK.  IntelliJ is using log4j internally but the plugin itself isn't using log4j and log4j is in the classpath because of the dependency on the SDK.
Do you know why this is needed atm? Is there another log4j dependency which needs to be moved to work with the SDK?

File: gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowGradlePlugin.kt (hunk 5 above)
Comment (i-walker) [reply gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowGradlePlugin.kt]: @jereksel Thanks for the explanation in Slack. The problem was that we were pointing to the gradle-plugin itself, which lead to the configurations we had back then to add the compiler-plugin manually.

File: compiler-plugin/build.gradle (hunk 4 above)
Comment (jereksel) [reply compiler-plugin/build.gradle]: I've removed dependency between Gradle and Kotlin compiler plugin. Maybe log4j was some transitive dependency.

In any case log4j that is used in tests is binary incompatible with one that is here.

File: gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowKotlinGradleSubplugin.kt (hunk 10 above)
Comment (jereksel) [reply gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowKotlinGradleSubplugin.kt]: So it'll be easier to add options in the future. I don't want to remove something that will surely be used in the future.

Comment (raulraja) [reply gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowKotlinGradleSubplugin.kt]: We need settings soon for all plugins and meta itself to enable and disable plugins via config without code changes so if this provides an entry path to it let's keep it.

File: README.md [hunk 12]
```diff
@@ -258,18 +258,36 @@ This is intentional so the API is all about intercepting desired elements and tr
...
 
-## Credits
+Quick start development environment:
 
-**Build and run tests**
+- Run `publishAndRunIde` Gradle task
+- Import sample project in new IDE instance
```
Comment (jereksel) [inline README.md]: I'll add it in another PR

File: gradle-plugin/build.gradle [hunk 13]
```diff
@@ -5,16 +5,13 @@ plugins {
 }
 apply plugin: 'kotlin'
 
-project.group = 'io.arrow-kt'
-project.version = '0.0.1'
```
Comment (rachelcarmena) [inline gradle-plugin/build.gradle]: Sorry @jereksel , I'm looking forward to merging this pull request because it has a lot of good things, thanks!, but I've seen these lines right now. I think it's necessary [a way to manage the group and version for Gradle plugin](https://plugins.gradle.org/docs/publish-plugin). As far as I know it cannot be the same process we are following so far. A Gradle plugin can only be published once with a specific version. However, snapshots can be published without limit. Thanks!

File: build.gradle [hunk 14]
```diff
@@ -87,13 +87,20 @@ subprojects { project ->
   }
 }
 
+task deployMeta {
+  dependsOn ':compiler-plugin:publishToMavenLocal'
+  dependsOn ':gradle-plugin:publishArrowPluginMarkerMavenPublicationToMavenLocal'
+  dependsOn ':gradle-plugin:publishPluginMavenPublicationToMavenLocal'
+  dependsOn ':gradle-plugin:publishToMavenLocal'
+}
```
Comment (rachelcarmena) [inline build.gradle]: It seems this task could be `installMeta` because it's publishing to the local repository.

File: gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowKotlinGradleSubplugin.kt [hunk 15]
```diff
@@ -13,9 +13,9 @@ import org.jetbrains.kotlin.gradle.plugin.SubpluginOption
  */
 class ArrowKotlinGradleSubplugin : KotlinGradleSubplugin<AbstractCompile> {
   companion object {
-    private const val GRADLE_ARTIFACT_NAME = "gradle-plugin"
+    private const val META_ARTIFACT_NAME = "compiler-plugin"
```
Comment (rachelcarmena) [inline gradle-plugin/src/main/kotlin/arrow/meta/plugin/gradle/ArrowKotlinGradleSubplugin.kt]: I checked this change when preparing #69. However, it seems that both `gradle-plugin` and `compiler-plugin` don't provide an additional feature for including the `compiler-plugin` with `-Xplugin`.
